node/block_exchange.go

```diff
@@ -0,0 +1,226 @@
+	conf         config.NodeConfig
+	genesis      *tmtypes.GenesisDoc
+	p2p          *p2p.Client
+	ex           *goheaderp2p.Exchange[*types.Block]
+	syncer       *goheadersync.Syncer[*types.Block]
+	sub          *goheaderp2p.Subscriber[*types.Block]
+	p2pServer    *goheaderp2p.ExchangeServer[*types.Block]
+	blockStore   *goheaderstore.Store[*types.Block]
+	syncerStatus *SyncerStatus
+
+	logger log.Logger
+	ctx    context.Context
+}
+
+func NewBlockExchangeService(ctx context.Context, store ds.TxnDatastore, conf config.NodeConfig, genesis *tmtypes.GenesisDoc, p2p *p2p.Client, logger log.Logger) (*HeaderExchangeService, error) {
+	// store is TxnDatastore, but we require Batching, hence the type assertion
+	// note, the badger datastore impl that is used in the background implements both
+	storeBatch, ok := store.(ds.Batching)
+	if !ok {
+		return nil, errors.New("failed to access the datastore")
+	}
+	ss, err := goheaderstore.NewStore[*types.SignedHeader](storeBatch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the header store: %w", err)
+	}
+
+	return &HeaderExchangeService{
+		conf:         conf,
+		genesis:      genesis,
+		p2p:          p2p,
+		ctx:          ctx,
+		headerStore:  ss,
+		logger:       logger,
+		syncerStatus: new(SyncerStatus),
+	}, nil
+}
+
+func (bExService *BlockExchangeService) initBlockStoreAndStartSyncer(ctx context.Context, initial *types.Block) error {
+	if err := bExService.blockStore.Init(ctx, initial); err != nil {
+		return err
+	}
+	if err := bExService.syncer.Start(bExService.ctx); err != nil {
+		return err
+	}
+	bExService.syncerStatus.m.Lock()
+	defer bExService.syncerStatus.m.Unlock()
+	bExService.syncerStatus.started = true
+	return nil
+}
+
+func (bExService *BlockExchangeService) tryInitBlockStoreAndStartSyncer(ctx context.Context, trustedBlock *types.Block) {
+	if trustedBlock != nil {
+		if err := bExService.initBlockStoreAndStartSyncer(ctx, trustedBlock); err != nil {
+			bExService.logger.Error("failed to initialize the headerstore and start syncer", "error", err)
+		}
+	}
+}
+
+func (bExService *BlockExchangeService) writeToBlockStoreAndBroadcast(ctx context.Context, signedHeader *types.SignedHeader) {
+	// For genesis header initialize the store and start the syncer
+	if signedHeader.Height() == bExService.genesis.InitialHeight {
+		if err := bExService.blockStore.Init(ctx, signedHeader); err != nil {
+			bExService.logger.Error("failed to initialize header store", "error", err)
+		}
+
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			bExService.logger.Error("failed to start syncer after initializing header store", "error", err)
+		}
+	}
+
+	// Broadcast for subscribers
+	if err := bExService.sub.Broadcast(ctx, signedHeader); err != nil {
+		bExService.logger.Error("failed to broadcast block header", "error", err)
+	}
+}
+
+// OnStart is a part of Service interface.
+func (bExService *BlockExchangeService) Start() error {
+	var err error
```

Comment:
nit: if you are define the error immediately, just do it in the return
```suggestion
func (bExService *BlockExchangeService) Start() (err error) {
```

Reply:
but also, looking at the rest of the method, I don't think you need to define it at all.  
So the line could just be deleted

```suggestion
func (bExService *BlockExchangeService) Start() error {
```